experimental/devices/mfrc522/mfrc522.go

```diff
@@ -0,0 +1,732 @@
+			return nil, err
+		}
+		if n&0x04 > 0 {
+			break
+		}
+	}
+	lsb, err := r.devRead(commands.CRCResultRegL)
+	if err != nil {
+		return nil, err
+	}
+
+	msb, err := r.devRead(commands.CRCResultRegM)
+	if err != nil {
+		return nil, err
+	}
+	return []byte{lsb, msb}, nil
+}
+
+// SelectTag selects the FOB device by device UUID.
+func (r *Dev) SelectTag(serial []byte) (byte, error) {
+	dataBuf := make([]byte, len(serial)+2)
+	dataBuf[0] = commands.PICC_SElECTTAG
+	dataBuf[1] = 0x70
+	copy(dataBuf[2:], serial)
+	crc, err := r.CRC(dataBuf)
+	if err != nil {
+		return 0, err
+	}
+	dataBuf = append(dataBuf, crc[0], crc[1])
+	backData, backLen, err := r.CardWrite(commands.PCD_TRANSCEIVE, dataBuf)
+	if err != nil {
+		return 0, err
+	}
+
+	var blocks byte
+
+	if backLen == 0x18 {
+		blocks = backData[0]
+	} else {
+		blocks = 0
+	}
+	return blocks, nil
+}
+
+// AuthStatus indicates the authentication response, could be one of AuthOk, AuthReadFailure or AuthFailure
+type AuthStatus byte
+
+const (
+	AuthOk AuthStatus = iota
+	AuthReadFailure
+	AuthFailure
+)
+
+func (r *Dev) auth(mode byte, blockAddress byte, sectorKey []byte, serial []byte) (AuthStatus, error) {
+	buffer := make([]byte, 2)
+	buffer[0] = mode
+	buffer[1] = blockAddress
+	buffer = append(buffer, sectorKey...)
+	buffer = append(buffer, serial[:4]...)
+	_, _, err := r.CardWrite(commands.PCD_AUTHENT, buffer)
+	if err != nil {
+		return AuthReadFailure, err
+	}
+	if n, err := r.devRead(commands.Status2Reg); err != nil || n&0x08 == 0 {
+		return AuthFailure, err
+	}
+	return AuthOk, nil
+}
+
+// StopCrypto stops the crypto chip.
+func (r *Dev) StopCrypto() error {
+	return r.clearBitmask(commands.Status2Reg, 0x08)
+}
+
+func (r *Dev) preAccess(blockAddr byte, cmd byte) ([]byte, int, error) {
+	send := make([]byte, 4)
+	send[0] = cmd
+	send[1] = blockAddr
+
+	crc, err := r.CRC(send[:2])
+	if err != nil {
+		return nil, -1, err
+	}
+	send[2] = crc[0]
+	send[3] = crc[1]
+	return r.CardWrite(commands.PCD_TRANSCEIVE, send)
+}
+
+func (r *Dev) read(blockAddr byte) ([]byte, error) {
+	data, _, err := r.preAccess(blockAddr, commands.PICC_READ)
+	if err != nil {
+		return nil, err
+	}
+	if len(data) != 16 {
+		return nil, fmt.Errorf("expected 16 bytes, actual %d", len(data))
+	}
+	return data, nil
+}
+
+func (r *Dev) write(blockAddr byte, data []byte) error {
+	read, backLen, err := r.preAccess(blockAddr, commands.PICC_WRITE)
+	if err != nil || backLen != 4 {
+		return err
+	}
+	if read[0]&0x0F != 0x0A {
+		return errors.New("can't authorize write")
+	}
+	newData := make([]byte, 18)
+	copy(newData, data[:16])
+	crc, err := r.CRC(newData[:16])
+	if err != nil {
+		return err
+	}
+	newData[16] = crc[0]
+	newData[17] = crc[1]
+	read, backLen, err = r.CardWrite(commands.PCD_TRANSCEIVE, newData)
+	if err != nil {
+		return err
+	}
+	if backLen != 4 || read[0]&0x0F != 0x0A {
+		err = errors.New("can not write data")
+	}
+	return nil
+}
+
+func calcBlockAddress(sector int, block int) (addr byte) {
```

Comment:
remove named return arguments everywhere, this code is just more opaque than the unnamed way.